### PR TITLE
Remove empty attribute parenthesis and attribute suffix if exists

### LINKED ISSUE #105: Removing empty attribute parentheses and Attribute suffix if present
Folks usually leave out the parentheses when there are no positional or named arguments:

```diff
-[Shouldly.ShouldlyMethodsAttribute()]
+[Shouldly.ShouldlyMethodsAttribute]
```

It's not common to see Attribute spelled out either and to me it feels noisy:

```diff
-[Shouldly.ShouldlyMethodsAttribute]
+[Shouldly.ShouldlyMethods]
```

What do you think?

### PR DESCRIPTION
**Pull request by danielmarbach**
fixes #105

### CODE REVIEW

File: src/PublicApiGenerator/ApiGenerator.cs [hunk 1]
```diff
@@ -191,6 +191,7 @@ static string NormaliseGeneratedCode(StringWriter writer)
             gennedClass = Regex.Replace(gennedClass, set, " { set; }", RegexOptions.IgnorePatternWhitespace);
             gennedClass = Regex.Replace(gennedClass, @"\s+{\s+}", " { }", RegexOptions.IgnorePatternWhitespace);
             gennedClass = Regex.Replace(gennedClass, @"\)\s+;", ");", RegexOptions.IgnorePatternWhitespace);
+            gennedClass = gennedClass.Replace("()]", "]"); // [Attribute] instead of [Attribute()]
```
Comment (jnm2) [inline src/PublicApiGenerator/ApiGenerator.cs]: Is there any need to defend against this replacement being done to a string literal?

Comment (danielmarbach) [reply src/PublicApiGenerator/ApiGenerator.cs]: I'm working on a slightly better way that also takes into account the attribute suffix replacement. I'm thinking about the following regex

```
@"\[(.*)Attribute\((.*)\)\]"
```

and then replace with

```
[$1]
```
if the second matching group is empty. 

What do you think about this?

Comment (jnm2) [reply src/PublicApiGenerator/ApiGenerator.cs]: That sounds good. It could still edit string literals, but the chance of it matching is getting smaller and smaller.

> if the second matching group is empty.

Then why not: `@"\[(.*)Attribute\(\)\]"`

Comment (sungam3r) [reply src/PublicApiGenerator/ApiGenerator.cs]: Looks good.

**Comment by danielmarbach** [comment]
I do think literals are escaped but I'll add a test case-------- Ursprüngliche Nachricht --------Betreff: Re: [ApiApprover/ApiApprover] Remove empty attribute parenthesis (#118)Von: Joseph Musser An: ApiApprover/ApiApprover Cc: Daniel Marbach ,Author @jnm2 commented on this pull request.



In src/PublicApiGenerator/ApiGenerator.cs:
> @@ -191,6 +191,7 @@ static string NormaliseGeneratedCode(StringWriter writer)
             gennedClass = Regex.Replace(gennedClass, set, " { set; }", RegexOptions.IgnorePatternWhitespace);
             gennedClass = Regex.Replace(gennedClass, @"\s+{\s+}", " { }", RegexOptions.IgnorePatternWhitespace);
             gennedClass = Regex.Replace(gennedClass, @"\)\s+;", ");", RegexOptions.IgnorePatternWhitespace);
+            gennedClass = gennedClass.Replace("()]", "]"); // [Attribute] instead of [Attribute()]

Is there any need to defend against this replacement being done to a string literal?

File: src/PublicApiGenerator/ApiGenerator.cs (hunk 1 above)
Comment (jnm2) [reply src/PublicApiGenerator/ApiGenerator.cs]: > Then why not: `@"\[(.*)Attribute\(\)\]"`

I wasn't thinking. You need both groups in order to remove `Attribute` when the parens aren't empty.

**Comment by danielmarbach** [comment]
The problem is nobody forces you to suffix attributes right? -------- Ursprüngliche Nachricht --------Betreff: Re: [ApiApprover/ApiApprover] Remove empty attribute parenthesis (#118)Von: Joseph Musser An: ApiApprover/ApiApprover Cc: Daniel Marbach ,Author @jnm2 commented on this pull request.



In src/PublicApiGenerator/ApiGenerator.cs:
> @@ -191,6 +191,7 @@ static string NormaliseGeneratedCode(StringWriter writer)
             gennedClass = Regex.Replace(gennedClass, set, " { set; }", RegexOptions.IgnorePatternWhitespace);
             gennedClass = Regex.Replace(gennedClass, @"\s+{\s+}", " { }", RegexOptions.IgnorePatternWhitespace);
             gennedClass = Regex.Replace(gennedClass, @"\)\s+;", ");", RegexOptions.IgnorePatternWhitespace);
+            gennedClass = gennedClass.Replace("()]", "]"); // [Attribute] instead of [Attribute()]

That sounds good. It could still edit string literals, but the chance of it matching is getting smaller and smaller.

if the second matching group is empty.

Then why not: @"\[(.*)Attribute\(\)\]"

**Comment by jnm2** [comment]
This one is going to be hard to explain, but it works in each combination of cases:

```cs
from input in new[]
{
	"[FooAttribute]",
	"[FooAttribute()]",
	"[FooAttribute(2)]",
	"[Foo]",
	"[Foo()]",
	"[Foo(2)]"
}
select Regex.Replace(input, @"(?<=\[.*)(Attribute)?(\(\)(?=\])|(?=\([^)]+\)\]))?", string.Empty)
```

Output:

```
[Foo]
[Foo]
[Foo(2)]
[Foo]
[Foo]
[Foo(2)]
```

**Comment by sungam3r** [comment]
Regular expressions have always been some sort magic to me.

**Comment by jnm2** [comment]
It's bad. I documented it and made some changes that seemed important while writing comments:

```cs
from input in new[]
{
    "[FooAttribute]",
    "[FooAttribute()]",
    "[FooAttribute(2)]",
    "[Foo]",
    "[Foo()]",
    "[Foo(2)]"
}
select Regex.Replace(
    input,
    @"
        (?<=\[[^\]]*)       # Require a starting square brace on the same line and any following characters except an ending square brace.
        (Attribute)?        # Delete this if present.
        (                   # Then require either:
            \(\)(?=\])      # - Empty parens (to delete) immediately followed by closing square brace (not deleted),
        |
            (?=\(.+\)\]) # - or non-empty parens immediately followed by closing square brace (not deleted).
        )",
    string.Empty,
    RegexOptions.IgnorePatternWhitespace)
```

**Comment by jnm2** [comment]
It's still pretty bad. Don't feel like you have to do anything with it. :D

**Comment by danielmarbach** [comment]
@sungam3r @jnm2 have a look. I think this should cover all the cases

**Comment by sungam3r** [comment]
What about this?
```c#
 public class Attribute1 : Attribute
 {
 }

 public class Attribute1Attribute : Attribute
 {
 }

 [@Attribute1Attribute]
 [@Attribute1]
 public class Class1
 {
 }
```

File: src/PublicApiGenerator/CodeNormalizer.cs [hunk 2]
```diff
@@ -33,6 +34,7 @@ public static string NormalizeGeneratedCode(StringWriter writer)
             gennedClass = Regex.Replace(gennedClass, Set, " { set; }", RegexOptions.IgnorePatternWhitespace);
             gennedClass = Regex.Replace(gennedClass, @"\s+{\s+}", " { }", RegexOptions.IgnorePatternWhitespace);
             gennedClass = Regex.Replace(gennedClass, @"\)\s+;", ");", RegexOptions.IgnorePatternWhitespace);
+            gennedClass = Regex.Replace(gennedClass, $@"(?s)(?<=\[[^\]]*)(Attribute)?({AttributeMarker}(\(\))?(?=\])|{AttributeMarker}(?=\(.+\)\])|{AttributeMarker}(?=new.+\}}\)\]))", string.Empty, RegexOptions.IgnorePatternWhitespace);
```
[the remark this answers is not in the document]
Comment (danielmarbach) [reply src/PublicApiGenerator/CodeNormalizer.cs]: Need multi line here for params object attributes. They cross multiple lines. If you remove it all those attribute test cases fail

[the remark this answers is not in the document]
Comment (danielmarbach) [reply src/PublicApiGenerator/CodeNormalizer.cs]: About the white space. There is. new object[] {...} also the params object attribute

[the remark this answers is not in the document]
Comment (jnm2) [reply src/PublicApiGenerator/CodeNormalizer.cs]: > Need multi line here for params object attributes. They cross multiple lines. If you remove it all those attribute test cases fail

My question is, why not use RegexOptions instead of putting it in the regex since you're already using `RegexOptions`?

```suggestion
            gennedClass = Regex.Replace(gennedClass, $@"(?<=\[[^\]]*)(Attribute)?({AttributeMarker}(\(\))?(?=\])|{AttributeMarker}(?=\(.+\)\])|{AttributeMarker}(?=new.+\}}\)\]))", string.Empty, RegexOptions.SingleLine | RegexOptions.IgnorePatternWhitespace);
```

[the remark this answers is not in the document]
Comment (jnm2) [reply src/PublicApiGenerator/CodeNormalizer.cs]: > About the white space. There is. new object[] {...} also the params object attribute

`RegexOptions.IgnorePatternWhitespace` means that whitespace in the pattern string is ignored. There is no whitespace in the string `$@"(?s)(?<=\[[^\]]*)(Attribute)?({AttributeMarker}(\(\))?(?=\])|{AttributeMarker}(?=\(.+\)\])|{AttributeMarker}(?=new.+\}}\)\]))"`.

You'd specify IgnorePatternWhitespace if you wanted to indent the regex pattern and leave comments explaining it, inside the pattern string.

[the remark this answers is not in the document]
Comment (danielmarbach) [reply src/PublicApiGenerator/CodeNormalizer.cs]: I should probably add an explanation for our future selfs

**Comment by danielmarbach** [comment]
Honestly, I have no good idea for now how to solve

https://github.com/ApiApprover/ApiApprover/pull/118#issuecomment-549905472

other than tracking all the attributes generated and then find out if a potential clash could be produced and then special name again. Which to me seems super silly for such an edge case.

And making everything pluggable and configurable seems overkill as well. Do you guys see this special case as a killer for this feature or should I proceed?

**Comment by jnm2** [comment]
Nope, I really don't think that situation will really happen in real-world projects. If it does, it won't be hard for the users to figure out what happened since their C# source code that they'll immediately compare with will look so unusual in the first place.

**Comment by sungam3r** [comment]
Nope, but let's leave a note about this special case near to the regex for attributes in `CodeNormalizer`.
